File: solutions/typescript/0084.ts
/**
 * 84. 柱状图中最大的矩形
 * 给定 n 个非负整数表示柱状图柱子高度（宽为 1），求能勾勒出的最大矩形面积。
 * 单调递增栈：栈中保存索引，对应高度严格递增。
 * 遇到 heights[i] < heights[stack.top] 时弹栈并以弹出的高度为基准计算面积：
 *   width = i - stack[top-1] - 1（栈空则为 i）。
 * 末尾追加 0 哨兵确保所有柱子被清算。
 */
function largestRectangleArea(heights: number[]): number {
    const stack: number[] = [];
    let maxArea = 0;
    const extended = heights.concat([0]);
    for (let i = 0; i < extended.length; i++) {
        while (stack.length > 0 && extended[stack[stack.length - 1]] > extended[i]) {
            const top = stack.pop()!;
            const h = extended[top];
            const w = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
            maxArea = Math.max(maxArea, h * w);
        }
        stack.push(i);
    }
    return maxArea;
}

export default largestRectangleArea;

File: solutions/rust/0084.rs
/// 84. 柱状图中最大的矩形
/// 单调递增栈：栈中保存索引，对应高度严格递增。
/// 遇到 heights[i] < heights[stack.top] 时弹栈并以弹出的高度为基准计算面积：
///   width = i - stack[stack.len()-1] - 1（栈空则为 i）。
/// 末尾追加 0 哨兵确保所有柱子被清算。
pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
    let n = heights.len();
    let mut extended: Vec<i32> = heights.clone();
    extended.push(0); // 哨兵

    let mut stack: Vec<usize> = Vec::new();
    let mut max_area: i32 = 0;

    for i in 0..extended.len() {
        let h = extended[i];
        while let Some(&top) = stack.last() {
            if extended[top] <= h {
                break;
            }
            stack.pop();
            let height = extended[top];
            let width = if stack.is_empty() {
                i as i32
            } else {
                (i - stack.last().unwrap() - 1) as i32
            };
            let area = height * width;
            if area > max_area {
                max_area = area;
            }
        }
        stack.push(i);
    }
    let _ = n;
    max_area
}

impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        largest_rectangle_area(heights)
    }
}